twine/package.py

```diff
@@ -46,8 +44,6 @@
     ".zip": "sdist",
 }
 
-MetadataValue = Union[str, Sequence[str], Tuple[str, IO, str]]
```

Comment:
Why are we getting less specific here?

Reply:
In https://github.com/pypa/twine/pull/608/commits/fa5ba3bb62d76c4e1061e04122b44990edf595fd, I used a `cast` on one of the dictionary values:

```python
    metadata = package.metadata_dictionary()
    description = metadata["description"]
    description_content_type = metadata["description_content_type"]

    # ...

    content_type, params = cgi.parse_header(cast(str, description_content_type))
```

Because `cgi.parse_header` expects a `str`, not the `Union[...]`. This led me to investigate how the `MetadataValue` alias is used. One place is `Repository._convert_data_to_list_of_tuples`:

https://github.com/pypa/twine/blob/7111240b04f022d5366af7297b9e81c63d76fb6d/twine/repository.py#L105-L108

However, to my eye, there's nothing about that code requires the strictness of `MetadataValue`.

Also, I've seen the `Dict[str, Any]` idiom [recommended by Guido](https://github.com/python/typing/issues/182#issuecomment-199532520) for dictionaries with mixed types(e.g. JSON) and I've used it in at least one other place:

https://github.com/pypa/twine/blob/8765d14d1ae54b35cdf28f2a5c76fdc6dc0f28c4/twine/repository.py#L74-L75

I think the ideal solution might be [`TypedDict`](https://docs.python.org/3/library/typing.html#typing.TypedDict) in Py 3.8 and/or type hints for `pkginfo`.

Reply:
That issue seems to be more tied to JSON than anything (by my skimming). I'd argue that `Any` is a barely valuable annotation and its main value is in "I haven't figured out what this should be right now". I don't understand how casting `description_content_type` to a string led us down this road, because that's exactly one of the values. 

I would also think that `pkginfo` is popular enough that there should already be some stubs for it, if not, it still seems like a valuable thing to have

Reply:
> I don't understand how casting `description_content_type` to a string led us down this road, because that's exactly one of the values.

I've been trying to avoid `cast` as a smell, though I have used it in a few places where mypy required it, e.g. when we know an `Optional[str]` is actually a `str`:

https://github.com/pypa/twine/blob/2cec2169faeb9a2906b2288776be91d660561a06/twine/commands/upload.py#L60

> I'd argue that `Any` is a barely valuable annotation and its main value is in "I haven't figured out what this should be right now".

I think it provides value to the reader indicating that "this value could be anything, so be careful". I also think how widely it's used is a matter of style, and it's more common when applying annotations to existing codebases.

https://mypy.readthedocs.io/en/stable/kinds_of_types.html#the-any-type
https://mypy.readthedocs.io/en/stable/casts.html

Do you feel strongly about this? I don't see much value in being so specific here, because it doesn't tell the reader or mypy enough to know what the runtime value of these dictionary keys will actually be. To me, that makes it just another thing that the reader has to look up.

> I would also think that `pkginfo` is popular enough that there should already be some stubs for it, if not, it still seems like a valuable thing to have

Looking into this (and other missing imports) is on the roadmap for subsequent PR's: https://github.com/pypa/twine/issues/231#issuecomment-619359423

Reply:
I also have used `cast` in a similar way in https://github.com/pypa/twine/blob/master/twine/commands/register.py#L25 while adding type annotations. 

I was wondering if assigning a value to `repository_url` if it is None might be a good idea rather than casting it.

```
 repository_url = register_settings.repository_config["repository"]
    if repository_url is None:
        repository_url = 'any'
```

I also saw a similar approach taking place in `twine/wininst.py`

https://github.com/pypa/twine/blob/d0ccb0c31e711a44f58a324ae65e1408f13ebee1/twine/wininst.py#L19-L25

Reply:
I don't feel strongly enough to block the PR, but I would like us to converge, eventually, on types that are of use to us and anyone else even if they seem to not convey they actual value of a dictionary. In general, I'd say something is better than nothing.

I'm also worried that we've released `MetadataValue` and people are using `twine.package` (which I think is a documented part of the API) and thus we're causing problems for others usin gmypy and our bare-bones annotations

Reply:
(One of these days I'll anticipate the API argument)

Just pushed https://github.com/pypa/twine/pull/608/commits/ebd2ada4a32f8db32d909aa7284fcb921f3e482c, which feels like a good compromise to me. I kept the relaxed signature of `_convert_data_to_list_of_tuples`:

https://github.com/pypa/twine/blob/ebd2ada4a32f8db32d909aa7284fcb921f3e482c/twine/repository.py#L106

That removed the need for the awkward `Tuple[str, IO, str]`, which is never actually stored in `metadata_dictionary`, but rather appended to the list of tuples:

https://github.com/pypa/twine/blob/ebd2ada4a32f8db32d909aa7284fcb921f3e482c/twine/repository.py#L152-L159

Leaving us with:

https://github.com/pypa/twine/blob/ebd2ada4a32f8db32d909aa7284fcb921f3e482c/twine/package.py#L48

So, yeah, glad I took a closer look at this. Thanks. 😉

Reply:
> I was wondering if assigning a value to `repository_url` if it is None might be a good idea rather than casting it. (I have assigned a value 'any', but I think any other standard value might be applicable here.

So we've had complaints from people that they've uploaded things they didn't intend to as well as to places they didn't intend to. Placing a standard value in `repository_url` to satisfy mypy doesn't seem safe to me. Sorry I didn't see your comment earlier, @deveshks

Reply:
@deveshks I think your use of `cast` is appropriate, and I've used it that way as well. 

Basically, given `my_dict: Dict[str, Optional[str]]`, with `cast(str, my_dict["key"])`, we're saying: "at this point in the code, I'm confident that a) "key" is in `my_dict` and b) it's value is a `str` and not `None`". Having that confidence requires us to reason about the logic up to that point, but in the cases I've seen so far, it's seemed safe (and sometimes obvious from the context).